Include/datetime.h

```diff
@@ -224,6 +209,8 @@ static PyDateTime_CAPI *PyDateTimeAPI = NULL;
 
 #define PyTZInfo_Check(op) PyObject_TypeCheck(op, PyDateTimeAPI->TZInfoType)
 #define PyTZInfo_CheckExact(op) (Py_TYPE(op) == PyDateTimeAPI->TZInfoType)
+#endif   /* !defined(_PY_DATETIME_IMPL) */
+
 
 /* Macros for accessing constructors in a simplified fashion. */
 #define PyDate_FromDate(year, month, day) \
```

Comment:
Macros like PyDate_FromDate() uses PyDateTimeAPI which is not defined if _PY_DATETIME_IMPL is defined. Maybe these macros should also be excluded if _PY_DATETIME_IMPL is defined?

Reply:
I don't think it matters very much. The macros under the `#ifndef` are the ones that get redefined in `_datetimemodules.c`. If you tried to use the other macros in `_datetimemodules.c`, they would fail to compile since they reference a variable that doesn't exist.

Reply:
> If you tried to use the other macros in _datetimemodules.c, they would fail to compile since they reference a variable that doesn't exist.

So maybe don't define them, no?

Reply:
Yeah, it's fine, I just don't think it matters one way or the other. I don't think it can cause problems and it won't affect the compiled output like having an unused static variable would.

I think it may be too much, but I'm sorta tempted to restructure `datetime.h` like this:

- `_datetime_common.h`: The first [180 lines of what is now `datetime.h`](https://github.com/python/cpython/blob/f624f90ef1dcf86389e87708934402751986f70e/Include/datetime.h#L3-L180), with the struct definitions in it
- `_datetime_api.h`: The [public facing API](https://github.com/python/cpython/blob/f624f90ef1dcf86389e87708934402751986f70e/Include/datetime.h#L189-L252)
- `datetime.h`:

```C
#ifndef DATETIME_H
#define DATETIME_H

#include "_datetime_common.h"
#include "_datetime_api.h"

#endif
```

Then `_datetimemodules.c` can include `_datetime_common.h`, and `datetime.h` doesn't need to know about the implementation details of `_datetimemodules.c`.

Reply:
Yeah, that's the other option. But I'm not really excited by this option. Just moving #endif sounds like a simpler solution ;-)